### PR TITLE
Bug 3053/v6

### PR DESCRIPTION
**Pull request by shivan1b**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/3053

### CODE REVIEW

File: src/app-layer-htp.c [hunk 1]
```diff
@@ -2640,7 +2641,15 @@ static void HTPConfigParseParameters(HTPCfgRec *cfg_prec, ConfNode *s,
                 cfg_prec->randomize = ConfValIsTrue(p->val);
             }
         } else if (strcasecmp("randomize-inspection-range", p->name) == 0) {
-            uint32_t range = atoi(p->val);
+            uint32_t range;
+            if (ByteExtractStringUint32(&range, 10, 0,
+                                        (const char *)p->val) < 0) {
+                FatalError(SC_ERR_FATAL, "Invalid value for randomize"
```
Comment (catenacyber) [inline src/app-layer-htp.c]: Why `FatalError` and not `SCLogError` as is done just below ?

File: src/detect-byte-extract.c [hunk 2]
```diff
@@ -305,7 +313,13 @@ static inline DetectByteExtractData *DetectByteExtractParse(const char *arg)
                            "for arg %d for byte_extract", i);
                 goto error;
             }
-            int multiplier = atoi(multiplier_str);
+            int multiplier;
```
Comment (catenacyber) [inline src/detect-byte-extract.c]: Should `multiplier` be declared as `int32` instead of `int` ?

File: src/detect-cipservice.c [hunk 3]
```diff
@@ -344,14 +344,18 @@ static DetectEnipCommandData *DetectEnipCommandParse(const char *rulestr)
         goto error;
     }
 
-    unsigned long cmd = atol(rulestr);
+    uint64_t cmd;
+    if (ByteExtractStringUint64(&cmd, 10, 0, rulestr) < 0) {
```
Comment (catenacyber) [inline src/detect-cipservice.c]: You should directly use `ByteExtractStringUint16` to avoid further cast

File: src/detect-engine-port.c [hunk 4]
```diff
@@ -1465,13 +1471,10 @@ DetectPort *PortParse(const char *str)
  */
 int DetectPortIsValidRange(char *port)
```
Comment (catenacyber) [inline src/detect-engine-port.c]: you could declare it static (so that we see that these changes do not affect other files)

File: src/detect-engine-port.c [hunk 5]
```diff
@@ -1413,21 +1414,25 @@ DetectPort *PortParse(const char *str)
...
+    if ((port2 = strchr(port, ':')) != NULL) {
         /* 80:81 range format */
         port2[0] = '\0';
         port2++;
 
-        if(DetectPortIsValidRange(port))
-            dp->port = atoi(port);
-        else
-            goto error;
+        if (strcmp(port, "") != 0) {
```
Comment (catenacyber) [inline src/detect-engine-port.c]: Why do we consider empty string means port 0 and not failure ?

File: src/detect-fast-pattern.c [hunk 6]
```diff
@@ -282,10 +283,11 @@ static int DetectFastPatternSetup(DetectEngineCtx *de_ctx, Signature *s, const c
                        "for fast_pattern offset");
             goto error;
         }
-        int offset = atoi(arg_substr);
-        if (offset > 65535) {
-            SCLogError(SC_ERR_INVALID_SIGNATURE, "Fast pattern offset exceeds "
-                       "limit");
+        int16_t offset;
+        if (ByteExtractStringInt16(&offset, 10, 0,
```
Comment (catenacyber) [inline src/detect-fast-pattern.c]: Should it rather be Uint16 ? 
Otherwise, it needs to be Int32 so that we can go up to 65535...

File: src/detect-ttl.c [hunk 7]
```diff
@@ -181,7 +182,11 @@ static DetectTtlData *DetectTtlParse (const char *ttlstr)
                     return NULL;
 
                 mode = DETECT_TTL_LT;
-                ttl1 = atoi(arg3);
+                if (ByteExtractStringInt32(&ttl1, 10, 0, (const char *)arg3) < 0) {
```
Comment (catenacyber) [inline src/detect-ttl.c]: IPv4 TTL is uint8 and not int32

But that may be part of another PR...

File: src/runmode-af-packet.c [hunk 8]
```diff
@@ -290,7 +295,10 @@ static void *ParseAFPConfig(const char *iface)
     if (ConfGetChildValueWithDefault(if_root, if_default, "cluster-id", &tmpclusterid) != 1) {
         aconf->cluster_id = (uint16_t)(cluster_id_auto++);
     } else {
-        aconf->cluster_id = (uint16_t)atoi(tmpclusterid);
+        if (ByteExtractStringInt32(&aconf->cluster_id, 10, 0, (const char *)tmpclusterid) < 0) {
```
Comment (catenacyber) [inline src/runmode-af-packet.c]: Uint16

File: src/util-proto-name.c [hunk 9]
```diff
@@ -57,10 +58,11 @@ void SCProtoNameInit()
             if (proto_ch == NULL)
                 continue;
 
-            int proto = atoi(proto_ch);
-            if (proto >= 255)
+            int proto;
+            if (ByteExtractStringInt32(&proto, 10, 0, (const char *)proto_ch) < 0)
```
Comment (catenacyber) [inline src/util-proto-name.c]: Uint8 ?

**Comment by jlucovsky** [comment]
Be consistent with value delimiters -- either `'` or `"` but not both.

File: src/util-ip.c [hunk 10]
```diff
@@ -65,7 +66,9 @@ bool IPv4AddressStringIsValid(const char *str)
 
     addr[dots][alen] = '\0';
     for (int x = 0; x < 4; x++) {
-        int a = atoi(addr[x]);
+        int a;
+        if (ByteExtractStringInt32(&a, 10, 0, (const char *)addr[x]) < 0)
+            return false;
```
Comment (jlucovsky) [inline src/util-ip.c]: Suggest including at least a debug message on failure.

File: src/util-napatech.c [hunk 11]
```diff
@@ -369,9 +369,12 @@ static uint32_t CountWorkerThreads(void)
 
                         char copystr[16];
                         strlcpy(copystr, lnode->val, 16);
-
-                        start = atoi(copystr);
-                        end = atoi(strchr(copystr, '-') + 1);
+                        if (ByteExtractStringUint8(&start, 10, 0, (const char *)copystr) < 0) {
+                            FatalError(SC_ERR_FATAL, "Napatech invalid worker range start");
```
Comment (jlucovsky) [inline src/util-napatech.c]: Suggest including the invalid value in the message.

File: src/util-proto-name.c (hunk 9 above)
Comment (jlucovsky) [reply src/util-proto-name.c]: I think that would be better -- `uint8`

File: src/app-layer-htp.c (hunk 1 above)
Comment (catenacyber) [reply src/app-layer-htp.c]: Ok, but I still think it is better to have only one oh those for both cases

Comment (jlucovsky) [reply src/app-layer-htp.c]: Agreed -- consistent use is important.

Comment (shivan1b) [reply src/app-layer-htp.c]: Thanks @catenacyber and @jlucovsky . This change was as per this discussion: https://github.com/OISF/suricata/pull/4169#discussion_r321640364 , Should I replace exit calls everywhere with FatalError? Probably that's what was meant by this comment.

File: src/detect-byte-extract.c (hunk 2 above)
Comment (shivan1b) [reply src/detect-byte-extract.c]: Yes, sure.

File: src/detect-cipservice.c (hunk 3 above)
Comment (shivan1b) [reply src/detect-cipservice.c]: Makes sense.

File: src/detect-engine-port.c (hunk 4 above)
Comment (shivan1b) [reply src/detect-engine-port.c]: Yes OK.

File: src/app-layer-htp.c (hunk 1 above)
Comment (victorjulien) [reply src/app-layer-htp.c]: Lets not further expand the scope of this work. Can you create a ticket for this instead?

File: src/detect-engine-port.c (hunk 5 above)
Comment (shivan1b) [reply src/detect-engine-port.c]: hmm I had thought that since port2's value in case it is empty, it is reset to 65535 then this should be reset to 0 since it is range of ports we're working on. So if we have `:5000`, it should probably be `0:5000`.
If this is incorrect which I believe it is from your comment, I would like to learn why.

File: src/detect-fast-pattern.c (hunk 6 above)
Comment (shivan1b) [reply src/detect-fast-pattern.c]: Yes OK

File: src/detect-dce-opnum.c [hunk 12]
```diff
@@ -175,16 +175,25 @@ static DetectDceOpnumData *DetectDceOpnumArgParse(const char *arg)
         if ((hyphen_token = index(dup_str_temp, '-')) != NULL) {
             hyphen_token[0] = '\0';
             hyphen_token++;
-            dor->range1 = atoi(dup_str_temp);
+            if (ByteExtractStringUint32(&dor->range1, 10, 0,
+                                        (const char *)dup_str_temp) < 0) {
+                goto error;
```
[the remark this answers is not in the document]
Comment (shivan1b) [reply src/detect-dce-opnum.c]: Hi Jeff!
Other goto error calls also did not log any messages so I left it as-is, should I add error messages in all the cases?

File: src/detect-engine.c [hunk 13]
```diff
@@ -2342,7 +2342,15 @@ static int DetectEngineCtxLoadConf(DetectEngineCtx *de_ctx)
             }
 
             if (insp_recursion_limit != NULL) {
-                de_ctx->inspection_recursion_limit = atoi(insp_recursion_limit);
+                if (ByteExtractStringInt32(&de_ctx->inspection_recursion_limit,
+                    10, 0, (const char *)insp_recursion_limit) < 0) {
+                    SCLogWarning(SC_ERR_INVALID_VALUE, "Invalid value for "
+                            "detect-engine.inspection-recursion-limit, "
```
[the remark this answers is not in the document]
Comment (shivan1b) [reply src/detect-engine.c]: Yes OK

File: src/runmode-pfring.c [hunk 14]
```diff
@@ -255,7 +263,11 @@ static void *ParsePfringConfig(const char *iface)
                 }
             }
         } else {
-            pfconf->threads = atoi(threadsstr);
+            if (ByteExtractStringUint16(&pfconf->threads, 10, 0, (const char *)threadsstr) < 0) {
+                SCLogWarning(SC_ERR_INVALID_VALUE, "Invalid value for "
+                             "threads: '%s'. Resetting to 1.", threadsstr);
```
[the remark this answers is not in the document]
Comment (shivan1b) [reply src/runmode-pfring.c]: Yes it should

File: src/util-cpu.c [hunk 15]
```diff
@@ -75,9 +76,12 @@ uint16_t UtilCpuGetNumProcessorsConfigured(void)
...
     return (uint16_t)nprocs;
 #elif OS_WIN32
-	long nprocs = -1;
-	const char* envvar = getenv("NUMBER_OF_PROCESSORS");
-	nprocs = (NULL != envvar) ? atoi(envvar) : 0;
+    int64_t nprocs = -1;
+    const char* envvar = getenv("NUMBER_OF_PROCESSORS");
+    if (envvar != NULL) {
+        if (ByteExtractStringInt64(&nprocs, 10, 0, envvar) < 0)
+            nprocs = 0;
```
[the remark this answers is not in the document]
Comment (shivan1b) [reply src/util-cpu.c]: Yes OK

File: src/app-layer-htp.c (hunk 1 above)
Comment (shivan1b) [reply src/app-layer-htp.c]: Yes sure. Here's the ticket: https://redmine.openinfosecfoundation.org/issues/3188

File: src/detect-engine-port.c (hunk 5 above)
Comment (catenacyber) [reply src/detect-engine-port.c]: Ok thanks, I did not know this was in a range.

Comment (catenacyber) [reply src/detect-engine-port.c]: So, should it be 0 or 0xFFFF ?

File: src/runmode-af-packet.c (hunk 8 above)
Comment (catenacyber) [reply src/runmode-af-packet.c]: If `cluster_id` is defined as an `int`, this means we have a double cast :
- first an explicit cast from `int` to `uint16` where 0x23456 becomes 0x3456 (and I am not sure what 0xFFFFFFFF becomes)
- then an implicit cast from `uint16` to `int` but the value does not change since every `uint16` value can be an `int`

If you use `ByteExtractStringUint16`, do you get a warning ?
Are there other places than this where `cluster_id` can have a value which is not an `uint16` (such as e negative value) ?
Otherwise, I would change the declaration of  `cluster_id` to be `uint16`

Comment (shivan1b) [reply src/runmode-af-packet.c]: > If `cluster_id` is defined as an `int`, this means we have a double cast :
> 
> * first an explicit cast from `int` to `uint16` where 0x23456 becomes 0x3456 (and I am not sure what 0xFFFFFFFF becomes)
> * then an implicit cast from `uint16` to `int` but the value does not change since every `uint16` value can be an `int`
> 
Thanks for the explanation.

> If you use `ByteExtractStringUint16`, do you get a warning ?
Yes. Since `ByteExtractStringUint16` expects its first argument to be `uint16_t` but instead I'm giving `int`.

```
runmode-af-packet.c: In function ‘ParseAFPConfig’:
runmode-af-packet.c:298:37: warning: passing argument 1 of ‘ByteExtractStringUint16’ from incompatible pointer type [-Wincompatible-pointer-types]
  298 |         if (ByteExtractStringUint16(&aconf->cluster_id, 10, 0, (const char *)tmpclusterid) < 0) {
      |                                     ^~~~~~~~~~~~~~~~~~
      |                                     |
      |                                     int *
In file included from runmode-af-packet.c:58:
util-byte.h:176:39: note: expected ‘uint16_t *’ {aka ‘short unsigned int *’} but argument is of type ‘int *’
  176 | int ByteExtractStringUint16(uint16_t *res, int base, uint16_t len, const char *str);
      |                             ~~~~~~~~~~^~~
```
> Are there other places than this where `cluster_id` can have a value which is not an `uint16` (such as e negative value) ?
There aren't any.

> Otherwise, I would change the declaration of `cluster_id` to be `uint16`
Done. Thanks a lot!

Comment (victorjulien) [reply src/runmode-af-packet.c]: Yes, this makes sense to do including for ```AFPThreadVars::cluster_id```. Please make sure to put these af-packet changes in their own commit.

File: src/detect-engine-port.c (hunk 5 above)
Comment (shivan1b) [reply src/detect-engine-port.c]: I think it should be 0 since strcmp returns 0 in case strings are equal. In this case, if start port is empty then strcmp would return 0.

Comment (catenacyber) [reply src/detect-engine-port.c]: Looks ok to me, thanks

File: src/detect-lua.c [hunk 16]
```diff
@@ -278,8 +279,16 @@ int DetectLuaMatchBuffer(DetectEngineThreadCtx *det_ctx,
                 SCLogDebug("k='%s', v='%s'", k, v);
 
                 if (strcmp(k, "retval") == 0) {
-                    if (atoi(v) == 1)
+                    int val;
+                    if (ByteExtractStringInt32(&val, 10, 0,
+                                               (const char *)v) < 0) {
+                        SCLogError(SC_ERR_INVALID_VALUE, "Invalid value in "
```
[the remark this answers is not in the document]
Comment (inashivb) [reply src/detect-lua.c]: Do you have a suggestion?

[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/detect-lua.c]: Maybe something like `"SCLogError(Invalid value for \"retval\" from LUA return table: '%s'", v)`

[the remark this answers is not in the document]
Comment (inashivb) [reply src/detect-lua.c]: Thanks, Jeff!

**Comment by inashivb** [comment]
This PR has been modified according to #4488 and broken into 4 parts which shall be coming along once 4488 is merged.
